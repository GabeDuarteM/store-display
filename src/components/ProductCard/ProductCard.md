```js
initialState = {
  isWishlisted: false,
}
;<ProductCard
  imgUrl="https://images.eastbay.com/pi/18837091/zoom/nike-legend-2.0-long-sleeve-t-shirt-mens"
  brand="Nike"
  id="5a5180507dceaccd20a314e0"
  name="Legend 2.0 Long Sleeve"
  price={24.99}
  isWishlisted={state.isWishlisted}
  isNew
  toggleIsWishlisted={() => setState({ isWishlisted: !state.isWishlisted })}
/>
```
