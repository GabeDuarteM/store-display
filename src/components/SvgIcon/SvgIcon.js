import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import cn from 'classnames'

export const styles = {
  svgIcon: {
    height: ({ height }) => height,
    width: ({ width }) => width,
    fill: ({ color }) => color,
    display: 'block',
  },
}

// eslint-disable-next-line no-unused-vars
const SvgIcon = ({
  classes,
  src,
  height,
  width,
  color,
  className,
  ...rest
}) => (
  <span
    className={cn(classes.svgIcon, className)}
    // eslint-disable-next-line react/no-danger
    dangerouslySetInnerHTML={{ __html: src }}
    {...rest}
  />
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** If passed, the className will be applied to the root of the component */
  className: PropTypes.string,
  /** A string containing the content of the svg */
  src: PropTypes.string.isRequired,
  /** The height of the svg */
  height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** The width of the svg */
  width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** The color to use as the fill of the svg */
  color: PropTypes.string,
}
export const defaultProps = {
  height: 'auto',
  width: 'auto',
  color: null,
  className: null,
}

SvgIcon.propTypes = propTypes
SvgIcon.defaultProps = defaultProps

export default injectSheet(styles)(SvgIcon)
