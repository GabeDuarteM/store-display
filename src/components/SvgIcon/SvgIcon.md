```js
const downArrow = require('../../assets/down-arrow.svg')
;<SvgIcon width={10} src={downArrow} />
```
