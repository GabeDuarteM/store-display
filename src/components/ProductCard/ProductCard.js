import React from 'react'

import PropTypes from 'prop-types'
import injectSheet from 'react-jss'
import cn from 'classnames'

import SvgIcon from '../SvgIcon/SvgIcon'
import heart from '../../assets/heart.svg'
import heartO from '../../assets/heart-o.svg'

export const styles = {
  productCard: {
    width: 280,
    height: 344,
    background: '#ffffff',
    overflow: 'hidden',
    transition: 'box-shadow 0.3s ease-in-out',
    '&:hover': {
      boxShadow: '0px 30px 50px rgba(0,0,0,0.15)',
    },
    '&:hover $wishlistButton': {
      visibility: 'visible !important',
    },
  },
  img: {
    backgroundImage: ({ imgUrl }) => `url(${imgUrl})`,
    backgroundPosition: 'center',
    backgroundSize: 'contain',
    backgroundRepeat: 'no-repeat',
    width: 260,
    height: 260,
    display: 'flex',
    justifyContent: 'space-between',
  },
  imgContainer: {
    width: 280,
    height: 280,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  wishlistButton: {
    visibility: 'hidden',
    cursor: 'pointer',
    transition: 'transform 180ms',
    transitionTimingFunction: 'cubic-bezier(.2,.7,.8,.7)',
    marginTop: 10,
    marginLeft: 8,
    '&:hover': {
      transform: 'scale(1.2,1.2)',
    },
    '&.wishlisted': {
      visibility: 'visible !important',
    },
  },
  newProduct: {
    width: 90,
    color: '#d0011b',
    height: 90,
    fontSize: 12,
    background: '#fff2f2',
    fontWeight: 500,
    display: 'flex',
    alignItems: 'flex-end',
    justifyContent: 'center',
    transform: 'rotate(45deg)',
    margin: -61,
    paddingBottom: 6,
    boxSizing: 'border-box',
    userSelect: 'none',
  },
  description: {
    display: 'flex',
    justifyContent: 'space-between',
    padding: 14,
  },
  brand: {
    color: '#585858',
    fontFamily: 'Montserrat',
    fontSize: 14.0,
    fontWeight: 500,
    textTransform: 'uppercase',
  },
  name: {
    color: '#646464',
    fontSize: 12.0,
    fontWeight: 300,
    width: 160,
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    marginTop: 4,
    height: 16,
  },
  price: {
    color: '#585858',
    fontSize: 14.0,
    fontWeight: 400,
    marginRight: 6,
  },
}

const ProductCard = ({
  classes,
  brand,
  name,
  price,
  id,
  currency,
  isWishlisted,
  toggleIsWishlisted,
  isNew,
  className,
}) => (
  <div className={cn(classes.productCard, className)}>
    <div className={classes.imgContainer}>
      <div className={classes.img}>
        <SvgIcon
          className={cn(classes.wishlistButton, { wishlisted: isWishlisted })}
          onClick={() => toggleIsWishlisted(id)}
          width={20}
          height={20}
          color={`#${isWishlisted ? 'ff5454' : '808080'}`}
          src={isWishlisted ? heart : heartO}
        />
        {isNew && <div className={classes.newProduct}>New</div>}
      </div>
    </div>
    <div className={classes.description}>
      <div>
        <div className={classes.brand}>{brand}</div>
        <div className={classes.name}>{name}</div>
      </div>
      <div className={classes.price}>
        {currency}
        {Intl.NumberFormat('pt-BR', { minimumFractionDigits: 2 }).format(price)}
      </div>
    </div>
  </div>
)

export const propTypes = {
  /** @ignore */
  classes: PropTypes.object.isRequired,
  /** The url of the image to be displayed */
  imgUrl: PropTypes.string.isRequired, // eslint-disable-line react/no-unused-prop-types
  /** The brand of the product */
  brand: PropTypes.string.isRequired,
  /** The name of the product */
  name: PropTypes.string.isRequired,
  /** The price of the product */
  price: PropTypes.number.isRequired,
  /** The id of the product */
  id: PropTypes.string.isRequired,
  /** Gets called when the user clicks the wishlist button */
  toggleIsWishlisted: PropTypes.func.isRequired,
  /** The currency of the product's price */
  currency: PropTypes.string,
  /** Prop that define if the product is wishlisted or not */
  isWishlisted: PropTypes.bool,
  /** Prop that define if the product is new or not */
  isNew: PropTypes.bool,
  /** custom className to be applied on the root element */
  className: PropTypes.string,
}

export const defaultProps = {
  currency: 'R$',
  isWishlisted: false,
  isNew: false,
  className: null,
}

ProductCard.propTypes = propTypes
ProductCard.defaultProps = defaultProps

export default injectSheet(styles)(ProductCard)
