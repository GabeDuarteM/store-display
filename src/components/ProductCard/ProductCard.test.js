import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import ProductCard, { styles, propTypes, defaultProps } from './ProductCard'
import SvgIcon from '../SvgIcon/SvgIcon'

const createWrapper = (func, optionalProps) =>
  func(
    <ProductCard
      imgUrl=""
      id="5a518050fbb2da0a6d228cf9"
      brand="Nike"
      name="Legend 2.0 Long Sleeve"
      price={24.99}
      toggleIsWishlisted={() => {}}
      {...optionalProps}
    />,
  )

const createWrapperShallow = optionalProps =>
  createWrapper(shallow, optionalProps)

const createWrapperMount = optionalProps => createWrapper(mount, optionalProps)

describe('COMPONENT: <ProductCard />', () => {
  it('should render without crashing', () => {
    createWrapperMount()
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it("should match component's snapshot with optional props", () => {
    const wrapper = createWrapperShallow({ isNew: true, isWishlisted: true })
    expect(toJson(wrapper.dive())).toMatchSnapshot()
  })

  it("should match component's style snapshot", () => {
    expect(styles).toMatchSnapshot()
  })

  it("should match component's propTypes snapshot", () => {
    expect(propTypes).toMatchSnapshot()
  })

  it("should match component's defaultProps snapshot", () => {
    expect(defaultProps).toMatchSnapshot()
  })

  it('should have an SvgIcon', () => {
    const wrapper = createWrapperShallow()

    const svgIconExists = wrapper
      .dive()
      .find(SvgIcon)
      .exists()
    expect(svgIconExists).toBe(true)
  })

  it('should have an SvgIcon', () => {
    const wrapper = createWrapperShallow()

    const svgIconExists = wrapper
      .dive()
      .find(SvgIcon)
      .exists()
    expect(svgIconExists).toBe(true)
  })

  it('should have an SvgIcon that calls toggleIsWishlisted when clicked', () => {
    const toggleIsWishlisted = jest.fn()
    const wrapper = createWrapperShallow({ toggleIsWishlisted })

    wrapper
      .dive()
      .find(SvgIcon)
      .simulate('click')
    expect(toggleIsWishlisted).toHaveBeenCalledTimes(1)
  })

  it('should have a red SvgIcon when isWishlisted is true', () => {
    const wrapper = createWrapperShallow({ isWishlisted: true })

    const color = wrapper
      .dive()
      .find(SvgIcon)
      .prop('color')
    expect(color).toBe('#ff5454')
  })

  it('should have a red SvgIcon when isWishlisted is true', () => {
    const wrapper = createWrapperShallow({ isWishlisted: false })

    const color = wrapper
      .dive()
      .find(SvgIcon)
      .prop('color')
    expect(color).toBe('#808080')
  })

  it('should have a wishlisted class when isWishlisted is true', () => {
    const wrapper = createWrapperShallow({ isWishlisted: true })

    const color = wrapper
      .dive()
      .find(SvgIcon)
      .hasClass('wishlisted')
    expect(color).toBe(true)
  })

  it('should not have a wishlisted class when isWishlisted is false', () => {
    const wrapper = createWrapperShallow({ isWishlisted: false })

    const color = wrapper
      .dive()
      .find(SvgIcon)
      .hasClass('wishlisted')
    expect(color).toBe(false)
  })

  it('should have a div with "New" as text when isNew is true', () => {
    const wrapper = createWrapperShallow({ isNew: true })

    const isNewDiv = wrapper
      .dive()
      .childAt(0)
      .childAt(0)
      .childAt(1)
    expect(isNewDiv.exists()).toBe(true)
    expect(isNewDiv.text()).toBe('New')
  })

  it('should not have a div with "New" as text when isNew is false', () => {
    const wrapper = createWrapperShallow({ isNew: false })

    expect(
      wrapper
        .dive()
        .childAt(0)
        .childAt(0)
        .children().length,
    ).toBe(1)
  })

  it('should correctly render the brand name', () => {
    const brand = 'brand name'
    const wrapper = createWrapperShallow({ brand })

    expect(
      wrapper
        .dive()
        .childAt(1)
        .childAt(0)
        .childAt(0)
        .text(),
    ).toBe(brand)
  })

  it('should correctly render the product price and currency', () => {
    const price = 15.99
    const currency = 'R$'
    const wrapper = createWrapperShallow({ price, currency })

    expect(
      wrapper
        .dive()
        .childAt(1)
        .childAt(1)
        .text(),
    ).toBe(currency + price)
  })
})
