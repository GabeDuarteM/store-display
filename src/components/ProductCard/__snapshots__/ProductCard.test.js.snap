// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <ProductCard /> should match component's defaultProps snapshot 1`] = `
Object {
  "className": null,
  "currency": "R$",
  "isNew": false,
  "isWishlisted": false,
}
`;

exports[`COMPONENT: <ProductCard /> should match component's propTypes snapshot 1`] = `
Object {
  "brand": [Function],
  "className": [Function],
  "classes": [Function],
  "currency": [Function],
  "id": [Function],
  "imgUrl": [Function],
  "isNew": [Function],
  "isWishlisted": [Function],
  "name": [Function],
  "price": [Function],
  "toggleIsWishlisted": [Function],
}
`;

exports[`COMPONENT: <ProductCard /> should match component's snapshot 1`] = `
<div
  className="ProductCard-productCard ProductCard-productCard"
>
  <div
    className="ProductCard-imgContainer ProductCard-imgContainer"
  >
    <div
      className="ProductCard-img ProductCard-img"
    >
      <Jss(SvgIcon)
        className="ProductCard-wishlistButton ProductCard-wishlistButton"
        color="#808080"
        height={20}
        onClick={[Function]}
        src=""
        width={20}
      />
    </div>
  </div>
  <div
    className="ProductCard-description ProductCard-description"
  >
    <div>
      <div
        className="ProductCard-brand ProductCard-brand"
      >
        Nike
      </div>
      <div
        className="ProductCard-name ProductCard-name"
      >
        Legend 2.0 Long Sleeve
      </div>
    </div>
    <div
      className="ProductCard-price ProductCard-price"
    >
      R$
      24.99
    </div>
  </div>
</div>
`;

exports[`COMPONENT: <ProductCard /> should match component's snapshot with optional props 1`] = `
<div
  className="ProductCard-productCard ProductCard-productCard"
>
  <div
    className="ProductCard-imgContainer ProductCard-imgContainer"
  >
    <div
      className="ProductCard-img ProductCard-img"
    >
      <Jss(SvgIcon)
        className="ProductCard-wishlistButton ProductCard-wishlistButton wishlisted"
        color="#ff5454"
        height={20}
        onClick={[Function]}
        src=""
        width={20}
      />
      <div
        className="ProductCard-newProduct ProductCard-newProduct"
      >
        New
      </div>
    </div>
  </div>
  <div
    className="ProductCard-description ProductCard-description"
  >
    <div>
      <div
        className="ProductCard-brand ProductCard-brand"
      >
        Nike
      </div>
      <div
        className="ProductCard-name ProductCard-name"
      >
        Legend 2.0 Long Sleeve
      </div>
    </div>
    <div
      className="ProductCard-price ProductCard-price"
    >
      R$
      24.99
    </div>
  </div>
</div>
`;

exports[`COMPONENT: <ProductCard /> should match component's style snapshot 1`] = `
Object {
  "brand": Object {
    "color": "#585858",
    "fontFamily": "Montserrat",
    "fontSize": 14,
    "fontWeight": 500,
    "textTransform": "uppercase",
  },
  "description": Object {
    "display": "flex",
    "justifyContent": "space-between",
    "padding": 14,
  },
  "img": Object {
    "backgroundImage": [Function],
    "backgroundPosition": "center",
    "backgroundRepeat": "no-repeat",
    "backgroundSize": "contain",
    "display": "flex",
    "height": 260,
    "justifyContent": "space-between",
    "width": 260,
  },
  "imgContainer": Object {
    "alignItems": "center",
    "display": "flex",
    "height": 280,
    "justifyContent": "center",
    "width": 280,
  },
  "name": Object {
    "color": "#646464",
    "fontSize": 12,
    "fontWeight": 300,
    "height": 16,
    "marginTop": 4,
    "overflow": "hidden",
    "textOverflow": "ellipsis",
    "whiteSpace": "nowrap",
    "width": 160,
  },
  "newProduct": Object {
    "alignItems": "flex-end",
    "background": "#fff2f2",
    "boxSizing": "border-box",
    "color": "#d0011b",
    "display": "flex",
    "fontSize": 12,
    "fontWeight": 500,
    "height": 90,
    "justifyContent": "center",
    "margin": -61,
    "paddingBottom": 6,
    "transform": "rotate(45deg)",
    "userSelect": "none",
    "width": 90,
  },
  "price": Object {
    "color": "#585858",
    "fontSize": 14,
    "fontWeight": 400,
    "marginRight": 6,
  },
  "productCard": Object {
    "&:hover": Object {
      "boxShadow": "0px 30px 50px rgba(0,0,0,0.15)",
    },
    "&:hover $wishlistButton": Object {
      "visibility": "visible !important",
    },
    "background": "#ffffff",
    "height": 344,
    "overflow": "hidden",
    "transition": "box-shadow 0.3s ease-in-out",
    "width": 280,
  },
  "wishlistButton": Object {
    "&.wishlisted": Object {
      "visibility": "visible !important",
    },
    "&:hover": Object {
      "transform": "scale(1.2,1.2)",
    },
    "cursor": "pointer",
    "marginLeft": 8,
    "marginTop": 10,
    "transition": "transform 180ms",
    "transitionTimingFunction": "cubic-bezier(.2,.7,.8,.7)",
    "visibility": "hidden",
  },
}
`;
